File: algorithm/stack/basic_calculator_ii.cpp
// https://leetcode.com/problems/basic-calculator-ii/

/*
Given a string s which represents an expression, evaluate this expression and return its value.
The integer division should truncate toward zero.
You may assume that the given expression is always valid. All intermediate results will be in the range of [-231, 231 - 1].
Note: You are not allowed to use any built-in function which evaluates strings as mathematical expressions, such as eval().

Ex1:
Input: s = "3+2*2"
Output: 7

Ex2:
Input: s = " 3/2 "
Output: 1

Ex3:
Input: s = " 3+5 / 2 "
Output: 5
*/

// 3+5 / 2 *4 -2

// [3,+,]
#include <vector>
#include <stack>
#include <string>
#include <iostream>

using namespace std;

int cal(int a, int b, char op) {
    if (op == '+') {
        return a+b;
    } else if (op == '-') {
        return a-b;
    } else if (op == '*') {
        return a*b;
    } else {
        return a/b;
    }
}

bool isdigit(char c) {
    return c >= '0' && c <= '9';
}

bool ismuldiv(char c) {
    return c == '*' || c == '/';
}

// Time: O(n), Space: O(n)
int calculate(string s) {
    if (s.empty()) {
        return 0;
    }

    vector<int> st;

    int n = s.size();
    int a = 0, b = 0;
    char c = '#';

    int i = 0;
    while (i < n) {
        if (s[i] == ' ') {
            i++;
            continue;
        }

        if (isdigit(s[i])) {
            string num = "";
            while (i < n && isdigit(s[i])) {
                num.push_back(s[i]);
                i++;
            }

            i--;

            if (c == '#') {
                st.push_back(stoi(num));
            } else {
                b = stoi(num);
                int res = cal(a, b, c);
                st.push_back(res);
            }
        } else if (ismuldiv(s[i])) {
            a = st.back();
            st.pop_back();
            c = s[i];
        } else {
            st.push_back(s[i]);
        }
        i++;
    }

    reverse(st.begin(), st.end());

    while (st.size() >= 3) {
        int b = st.back();
        st.pop_back();
        char c = st.back();
        st.pop_back();
        int a = st.back();
        st.pop_back();

        cout << "a: " << a << " b:" << b << endl;
        int res = cal(a, b, c);
        st.push_back(res);
    }

    return st.back();
}

int main() {
    // cout << calculate("3+2*2") << endl; // 7
    // cout << calculate(" 3/2 ") << endl; // 1
    // cout << calculate(" 3+5 / 2") << endl; // 5

    cout << calculate("1-1+1") << endl; // -1
    return 0;
}

File: algorithm/stack/python/basic_calculator_ii.py
"""
Basic Calculator II
https://leetcode.com/problems/basic-calculator-ii/

Given a string s which represents an expression, evaluate this expression
and return its value. Integer division truncates toward zero.

Approaches:
    1. calculate  - Stack-based evaluation. Time: O(n), Space: O(n).
    2. calculate2 - Running result with two accumulators. Time: O(n), Space: O(1).
"""


# Time: O(n), Space: O(n)
def calculate(s: str) -> int:
    res = 0
    num = 0
    n = len(s)

    op = '+'
    st = []
    for i in range(n):
        c = s[i]
        if '0' <= c <= '9':
            num = num * 10 + (ord(c) - ord('0'))
        if i == n - 1 or (c < '0' and c != ' '):
            if op == '+':
                st.append(num)
            elif op == '-':
                st.append(-num)
            else:
                top = st.pop()
                if op == '*':
                    tmp = top * num
                else:
                    # Truncate toward zero (C++ integer division semantics)
                    tmp = int(top / num)
                st.append(tmp)
            op = c
            num = 0

    while st:
        res += st.pop()

    return res


# Time: O(n), Space: O(1)
def calculate2(s: str) -> int:
    res = 0
    cur_res = 0
    num = 0
    n = len(s)
    op = '+'
    for i in range(n):
        c = s[i]
        if '0' <= c <= '9':
            num = num * 10 + ord(c) - ord('0')
        if c in ('+', '-', '*', '/') or i == n - 1:
            if op == '+':
                cur_res += num
            elif op == '-':
                cur_res -= num
            elif op == '*':
                cur_res *= num
            elif op == '/':
                # Truncate toward zero (C++ integer division semantics)
                cur_res = int(cur_res / num)
            if c in ('+', '-') or i == n - 1:
                res += cur_res
                cur_res = 0
            op = c
            num = 0
    return res


if __name__ == "__main__":
    print(calculate2("3+2*2"))      # 7
    print(calculate2(" 3/2 "))      # 1
    print(calculate2(" 3+5 / 2"))   # 5
    print(calculate2("1-1+1"))      # 1
